zcash_client_memory/src/types/account.rs

```diff
@@ -0,0 +1,780 @@
+                    }
+                }
+                Ok(None)
+            }
+        }
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    pub(crate) fn find_account_for_ephemeral_address(
+        &self,
+        address: &TransparentAddress,
+    ) -> Result<Option<AccountId>, Error> {
+        for (account_id, account) in self.accounts.iter() {
+            let contains = account
+                .ephemeral_addresses()?
+                .iter()
+                .any(|(eph_addr, _)| eph_addr == address);
+            if contains {
+                return Ok(Some(*account_id));
+            }
+        }
+        Ok(None)
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    pub(crate) fn mark_ephemeral_address_as_seen(
+        &mut self,
+        address: &TransparentAddress,
+        tx_id: TxId,
+    ) -> Result<(), Error> {
+        for (_, account) in self.accounts.iter_mut() {
+            account.mark_ephemeral_address_as_seen(address, tx_id)?
+        }
+        Ok(())
+    }
+}
+
+impl Deref for Accounts {
+    type Target = BTreeMap<AccountId, Account>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.accounts
+    }
+}
+
+impl DerefMut for Accounts {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.accounts
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct EphemeralAddress {
+    pub(crate) address: TransparentAddress,
+    // Used implies seen
+    pub(crate) used: Option<TxId>,
+    pub(crate) seen: Option<TxId>,
+}
+
+impl EphemeralAddress {
+    fn mark_used(&mut self, tx: TxId) {
+        // We update both `used_in_tx` and `seen_in_tx` here, because a used address has
+        // necessarily been seen in a transaction. We will not treat this as extending the
+        // range of addresses that are safe to reserve unless and until the transaction is
+        // observed as mined.
+        self.used.replace(tx);
+        self.seen.replace(tx);
+    }
+    fn mark_seen(&mut self, tx: TxId) -> Option<TxId> {
+        self.seen.replace(tx)
+    }
+}
+
+/// An internal representation account stored in the database.
+#[derive(Debug, Clone)]
+pub struct Account {
+    account_name: String,
+    account_id: AccountId,
+    kind: AccountSource,
+    viewing_key: UnifiedFullViewingKey,
+    birthday: AccountBirthday,
+    /// Stores diversified Unified Addresses that have been generated from accounts in the wallet.
+    addresses: BTreeMap<DiversifierIndex, UnifiedAddress>,
+    pub(crate) ephemeral_addresses: BTreeMap<u32, EphemeralAddress>, // NonHardenedChildIndex (< 1 << 31)
+    _notes: BTreeSet<NoteId>,
+}
+
+impl PartialEq for Account {
+    fn eq(&self, other: &Self) -> bool {
+        self.account_name == other.account_name
+            && self.account_id == other.account_id
+            && self.kind == other.kind
+            && self
+                .viewing_key
+                .encode(&zcash_primitives::consensus::MainNetwork)
```

Comment:
[non-blocking] 
is this because `UnifiedFullViewingKey` does not implement Eq?  Isn't this a big performance hit in terms of what it seems to be an innocuous comparison that covertly does two encoding operations?

@nuttycom what's the proper way of telling if two UFVKs are the same?